app/ghcup/BrickMain.hs

```diff
@@ -59,70 +74,243 @@ import qualified Data.Text.Lazy.Builder        as B
+handleGenericListEvent (MouseDown _ Vty.BScrollDown _ _)          = moveDown
+handleGenericListEvent (MouseDown _ Vty.BScrollUp _ _)            = moveUp
+handleGenericListEvent (VtyEvent (Vty.EvKey Vty.KDown []))     = moveDown
+handleGenericListEvent (VtyEvent (Vty.EvKey Vty.KUp []))       = moveUp
+handleGenericListEvent (VtyEvent ev) = do
+    ring <- use sectionListFocusRingL
+    case F.focusGetCurrent ring of
+        Nothing -> pure ()
+        Just l  -> Brick.zoom (toLensVL $ sectionL l) $ L.handleListEvent ev
+handleGenericListEvent _ = pure ()
+
+-- This re-uses Brick.Widget.List.renderList
+renderSectionList :: (Traversable t, Ord n, Show n, Eq n, L.Splittable t)
+                  => (Bool -> n -> Widget n)             -- ^ Rendering function for separator between sections. True for selected before section
+                  -> (Bool -> n -> Widget n -> Widget n) -- ^ Rendering function for the borders. True for selected section
+                  -> (Bool -> e -> Widget n)             -- ^ Rendering function of the list element, True for the selected element
+                  -> Bool                                -- ^ Whether the section list has focus
+                  -> GenericSectionList n t e            -- ^ The section list to render
+                  -> Widget n
+renderSectionList render_separator render_border render_elem section_focus (GenericSectionList focus elms sl_name) =
+  Widget Greedy Greedy $ render $
+    Brick.viewport sl_name Vertical $
+      V.foldl' (\wacc list ->
+                  let has_focus_list = is_focused_section list
+                      list_name = L.listName list
+                  in   wacc
+                    <=> render_separator has_focus_list list_name
+                    <=> inner_widget has_focus_list list_name list
+              )
+              Brick.emptyWidget elms
+  where
+    is_focused_section l = section_focus && Just (L.listName l) == F.focusGetCurrent focus
+    inner_widget has_focus k l = 
+      let w = render_border has_focus k (Brick.vLimit (length l) $ L.renderList render_elem has_focus l)
+       in if has_focus then Brick.visible w else w
 
 
-hiddenTools :: [Tool]
-hiddenTools = []
+-- | Equivalent to listSelectedElement
+sectionListSelectedElement :: (Eq n, L.Splittable t, Traversable t, Semigroup (t e)) => GenericSectionList n t e -> Maybe (Int, e)
+sectionListSelectedElement generic_section_list = do
+  current_focus <- generic_section_list ^. sectionListFocusRingL & F.focusGetCurrent 
+  let current_section = generic_section_list ^. sectionL current_focus
+  L.listSelectedElement current_section 
+
+{- GHCUp State
+
+-}
+
+data Name = AllTools       -- The main list widget
+          | Singular Tool  -- The particular list for each tool
+          | IODialog       -- The pop up when installing a tool
+          | TutorialBox    -- The tutorial widget
+          deriving (Eq, Ord, Show)
 
+data Mode = Navigation | Tutorial deriving (Eq, Show, Ord)
+
```

Comment:
The `Mode` data type is the [way to go](https://github.com/jtdaugherty/brick/issues/476#issuecomment-1629151920) for `Brick` applications. It isn't part of the fundamentals of brick, but any complex example use it as a sort of "state machine" type